File: test/end-to-end/VectorClass/Ruby/vector.rb
require "set"
class Vector
    attr_accessor :data
    attr_accessor :capacity
    attr_accessor :length

    def initialize()
        self.capacity = 0
        self.data = []
        self.length = 0
    end

    def at(index)
        if index >= self.length
            raise Exception.new("Index out of bounds: %d is greater than %d." % [index, self.length])
        end

        return self.data[index]
    end

    def get_capacity()
        return self.capacity
    end

    def get_length()
        return self.length
    end

    def get_first()
        if self.capacity == 0
            raise Exception.new("Cannot get first from empty vector.")
        end

        return self.data[0]
    end

    def get_last()
        if self.capacity == 0
            raise Exception.new("Cannot get last from empty vector.")
        end

        return self.data[self.length - 1]
    end

    def ensure_capacity(capacity)
        if capacity <= self.capacity
            return
        end

        newCapacity = (capacity / 2.0).ceil * 3.floor
        oldData = self.data
        self.capacity = newCapacity
        self.data = Array.new(newCapacity)

        for i in 0...oldData.length
            self.data[i] = oldData[i]
        end
    end

    def push(item)
        self.ensure_capacity(self.length + 1)

        self.data[self.length] = item
        self.length += 1
    end

    def resize(length)
        if length <= self.length
            self.length = length
            return
        end

        self.ensure_capacity(length)
        self.length = length
    end

    def to_array()
        array = Array.new(self.length)

        for i in 0...self.length
            array[i] = self.data[i]
        end

        return array
    end

    def to_list()
        list = Array.new(self.length)

        for i in 0...self.length
            list[i] = self.data[i]
        end

        return list
    end

    def to_set()
        individuals = Set.new

        for i in 0...self.length
            individuals.add(self.data[i])
        end

        return individuals
    end
end

File: test/end-to-end/VectorClass/TypeScript/Vector.ts
export class Vector<T> {
    private data: T[];
    private capacity: number;
    private length: number;

    public constructor() {
        this.capacity = 0;
        this.data = [];
        this.length = 0;
    }

    public at(index: number): T {
        if (index >= this.length) {
            throw new Error(`Index out of bounds: ${index} is greater than ${this.length}.`);
        }

        return this.data[index];
    }

    public getCapacity(): number {
        return this.capacity;
    }

    public getLength(): number {
        return this.length;
    }

    public getFirst(): T {
        if (this.capacity === 0) {
            throw new Error("Cannot get first from empty vector.");
        }

        return this.data[0];
    }

    public getLast(): T {
        if (this.capacity === 0) {
            throw new Error("Cannot get last from empty vector.");
        }

        return this.data[this.length - 1];
    }

    public ensureCapacity(capacity: number): void {
        if (capacity <= this.capacity) {
            return;
        }

        let newCapacity: number = Math.floor(Math.ceil((capacity / 2.0)) * 3);
        let oldData: T[] = this.data;
        this.capacity = newCapacity;
        this.data = new Array<T>(newCapacity);

        for (let i: number = 0; i < oldData.length; i += 1) {
            this.data[i] = oldData[i];
        }
    }

    public push(item: T): void {
        this.ensureCapacity(this.length + 1);

        this.data[this.length] = item;
        this.length += 1;
    }

    public resize(length: number): void {
        if (length <= this.length) {
            this.length = length;
            return;
        }

        this.ensureCapacity(length);
        this.length = length;
    }

    public toArray(): T[] {
        let array: T[] = new Array<T>(this.length);

        for (let i: number = 0; i < this.length; i += 1) {
            array[i] = this.data[i];
        }

        return array;
    }

    public toList(): T[] {
        let list: T[] = new Array<T>(this.length);

        for (let i: number = 0; i < this.length; i += 1) {
            list[i] = this.data[i];
        }

        return list;
    }

    public toSet(): Set<T> {
        let individuals: Set<T> = new Set<T>();

        for (let i: number = 0; i < this.length; i += 1) {
            individuals.add(this.data[i]);
        }

        return individuals;
    }
}
